test/tests.hs

```diff
@@ -107,31 +110,78 @@ main = defaultMain $ testGroup "discrimination"
     , testSorting (Proxy :: Proxy Char)
     , testSorting (Proxy :: Proxy String)
     , testSorting (Proxy :: Proxy (NonEmpty Int))
+    , testSorting (Proxy :: Proxy Natural)
+    , testSorting (Proxy :: Proxy Integer)
+
+    , testSorting' (\xs -> L.foldl' (*) 1 (map fromIntegral (xs :: [Word64])) :: Natural)
+    , testSorting' (\xs -> L.foldl' (*) 1 (map fromIntegral (xs :: [Int64]))  :: Integer)
     ]
   ]
 
 testGrouping
   :: forall a. (Grouping a, Typeable a, Arbitrary a, Eq a, Show a)
   => Proxy a
   -> TestTree
-testGrouping p = testProperty name prop
-  where
-    name = show (typeRep p)
+testGrouping _ = testGrouping' (id :: a -> a)
 
-    prop :: a -> a -> Property
-    prop x y = label (show lhs) $ lhs === groupingEq x y
+testGrouping'
+  :: forall a b. (Grouping b, Typeable a, Typeable b, Arbitrary a, Eq b, Show a, Show b)
+  => (a -> b)
+  -> TestTree
+testGrouping' f = testGroup name
+    [ testProperty "groupingEq" prop_eq
+    , testProperty "nub"        prop_nub
+    ]
+  where
+    tra = typeRep (Proxy :: Proxy a)
+    trb = typeRep (Proxy :: Proxy b)
+    name = if tra == trb then show tra else show trb ++ " from " ++ show tra
+
+    prop_eq :: a -> a -> Property
+    prop_eq x' y' =
+        counterexample (show (x,y)) $
+        label (show lhs) $
+        lhs === groupingEq x y
```

Comment:
Not objecting or anything, but just curious what `counterexample` and `label` add over what `===` already provides?

Reply:
`label` shows the distribution, i.e. even in successful runs we see the percentages:

```
   Natural from [Word64]
      sortingCompare: OK (0.11s)
        +++ OK, passed 100 tests:
        48% GT
        48% LT
         4% EQ
```

(I suspect that `EQ` cases are small though).

`counterexample` shows the `x` and `y`, as `===` (and `property`) would show only the initial inputs and sides of `===`, not intermediate values. E.g.

```
        [1,0,0,0,0,0,0,0,0,0,0,1]
        [1,0,0,0,0,0,0,0,0,0,1,0]
                [1,0,0,0,0,0,0,0,0,0,0,1]
        [1,0,0,0,0,0,0,0,0,0,1,0]
        (84162174424773976115855838126082058648805436845170781751972494449099714468753293153818664580441415219631727501698851483408310916002940861810045036330430093599283578738055113571066620126149163476340692027772502017,84162174424773976115855838126082058648805436845170781751972494449099714468753293153818664580441415219631727501698851483408310916002940861810045036330430093599283578738055113571066620126149163494787436101482053632)
        LT /= GT
```

(shrinking almost works).